snapcraft/plugins/maven.py

```diff
@@ -64,39 +90,108 @@ def schema(cls):
             "default": [""],
         }
 
+        schema["properties"]["maven-version"] = {"type": "string"}
+
+        schema["properties"]["maven-version-checksum"] = {"type": "string"}
+
+        schema["required"] = ["source"]
+        schema["properties"]["maven-openjdk-version"] = {
+            "type": "string",
+            "default": "",
+        }
         schema["required"] = ["source"]
 
         return schema
 
-    def __init__(self, name, options, project):
-        super().__init__(name, options, project)
-        self.build_packages.append("maven")
-
-    def _use_proxy(self):
-        return any(k in os.environ for k in ("http_proxy", "https_proxy"))
+    @classmethod
+    def get_pull_properties(cls):
+        # Inform Snapcraft of the properties associated with pulling. If these
+        # change in the YAML Snapcraft will consider the pull step dirty.
+        return ["maven-version", "maven-version-checksum", "maven-openjdk-version"]
 
     @classmethod
     def get_build_properties(cls):
         # Inform Snapcraft of the properties associated with building. If these
         # change in the YAML Snapcraft will consider the build step dirty.
         return ["maven-options", "maven-targets"]
 
+    @property
+    def _maven_tar(self):
+        if self._maven_tar_handle is None:
+            maven_uri = _MAVEN_URL.format(version=self._maven_version)
+            self._maven_tar_handle = sources.Tar(
+                maven_uri, self._maven_dir, source_checksum=self._maven_checksum
+            )
+        return self._maven_tar_handle
+
+    def __init__(self, name, options, project):
+        super().__init__(name, options, project)
+
+        self._setup_maven()
+        self._setup_base_tools(project.info.base)
+
+    def _setup_base_tools(self, base):
+        version = self.options.maven_openjdk_version
+        if base == "core16":
+            if not version:
+                version = "8"
```

Comment:
Why 8 and not 9?

Reply:
I was trying something out, let me move that out of the way.